packages/collapsible-side-nav/src/SideNav.tsx

```diff
@@ -0,0 +1,204 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import debounce from 'lodash/debounce';
+import { transparentize } from 'polished';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { useEventListener } from '@leafygreen-ui/hooks';
+import { HTMLElementProps, IdAllocator } from '@leafygreen-ui/lib';
+import { uiColors } from '@leafygreen-ui/palette';
+import CollapseButton from './CollapseButton';
+import { SideNavContext } from './contexts';
+import { transitionDurationMilliseconds } from './utils';
+
+const navStyle = css`
+  position: relative;
+  z-index: 0;
+  min-width: 184px;
+  background-color: ${uiColors.gray.light3};
+  border-right: 1px solid ${uiColors.gray.light2};
+  transition: min-width ${transitionDurationMilliseconds}ms ease-in-out,
+    box-shadow ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavStyle = css`
+  position: absolute;
+  min-width: 48px;
+  width: 48px;
+  top: 0;
+  bottom: 0;
+`;
+
+const collapsedHoveredNavStyle = css`
+  position: absolute;
+  top: 0;
+  bottom: 0;
+  border-width: 0;
+  box-shadow: 6px 0 10px -8px ${transparentize(0.7, uiColors.black)};
+`;
+
+const collapsedSpacerStyle = css`
+  min-width: 48px;
+  width: 48px;
+`;
+
+const navListStyle = css`
+  position: relative;
+  padding-top: 12px;
+  height: 100%;
+  overflow: hidden;
+  transition: all ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavListStyle = css`
+  padding-top: 64px;
+`;
+
+type Props = {
+  children: React.ReactNode;
+  className?: string;
+  currentPath?: string;
+} & (
+  | { collapsible: false; initialCollapsed?: undefined }
+  | { collapsible?: true; initialCollapsed?: boolean }
+) &
+  Omit<HTMLElementProps<'nav'>, 'ref'>;
+
+const idAllocator = IdAllocator.create('side-nav');
+
+const SideNav = React.forwardRef<HTMLElement, Props>(function SideNav(
+  {
+    children,
+    className,
+    currentPath,
+    collapsible = true,
+    initialCollapsed = false,
+    ...rest
+  }: Props,
+  ref,
+) {
+  const id = useMemo(() => idAllocator.generate(), []);
+
+  const [collapsed, setCollapsed] = useState(initialCollapsed);
+
+  // Manually detect hover so that hover state can be passed through the context.
+  // Using pure CSS would require the nav component to import a data prop selector
+  // from every sub-component or an equivalent amount of work.
+  const [hovered, setHovered] = useState(false);
+
+  // Debounce hover changes to prevent jitter when rapidly moving the mouse
+  // on and off the nav. Alternatively we could set transition-delay for this
+  // and all sub-components, but that would be less convenient and wouldn't
+  // prevent unnecessary re-renders.
+  const debouncedSetHovered = useMemo(
+    () => ({
+      toTrue: debounce(() => setHovered(true), transitionDurationMilliseconds),
```

Comment:
The transition in still feels a little jittery to me, while the transition out does not